src/shared/dissect-image.c

```diff
@@ -1289,8 +1295,37 @@ static int verity_partition(
 #endif
         } else
                 r = crypt_activate_by_volume_key(cd, name, root_hash, root_hash_size, CRYPT_ACTIVATE_READONLY);
-        if (r < 0)
+        if (r < 0 && r != -EEXIST)
                 return r;
+        if (r == -EEXIST) {
+                /* If the same volume was already open, check that the root hashes match, and reuse it if they do */
+                _cleanup_free_ char *root_hash_existing = NULL;
+                struct crypt_params_verity crypt_params = {};
+                size_t root_hash_existing_size = root_hash_size;
+
+                crypt_free(cd);
+                r = crypt_init_by_name(&cd, name);
+                if (r < 0)
+                        return log_error_errno(r, "Error opening verity device, crypt_init_by_name failed: %m");
+                r = crypt_get_verity_info(cd, &crypt_params);
+                if (r < 0)
+                        return log_error_errno(r, "Error opening verity device, crypt_get_verity_info failed: %m");
+                root_hash_existing = malloc0(root_hash_size);
+                if (!root_hash_existing)
+                        return log_oom();
+                r = crypt_volume_key_get(cd, CRYPT_ANY_SLOT, root_hash_existing, &root_hash_existing_size, NULL, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error opening verity device, crypt_volume_key_get failed: %m");
+                if (memcmp(root_hash_existing, root_hash, root_hash_size))
+                        return log_error_errno(SYNTHETIC_ERRNO(-EINVAL), "Error opening verity device, it already exists but root hashes are different.");
+#if HAVE_CRYPT_ACTIVATE_BY_SIGNED_KEY
+                /* Ensure that, if signatures are supported, we only reuse the device if the previous mount
+                 * used the same settings, so that a previous unsigned mount will not be reused if the user
+                 * asks to use signing for the new one, and viceversa. */
+                if ((!!root_hash_sig || !!root_hash_sig_path) != !!(crypt_params.flags & CRYPT_VERITY_ROOT_HASH_SIGNATURE))
+                        return log_error_errno(SYNTHETIC_ERRNO(-EINVAL), "Error opening verity device, it already exists but signature settings are not the same.");
+#endif
+        }
```

Comment:
hmm, so this is going to be racy. If you are much more aggressively sharing DM devices here, then if we stop a service that is backed by a DM device while you start another one backed by the same you will likely run into issues here...

I figure, if aggressive sharing of DM devices is the goal here this needs to be made robust. i.e. the above needs to be a loop: try to allocate the device, and if that fails with EEXIST try to open the existing one. if that fails with ENOENT/ENODEV, then try to allocate it anew again and so on. Also, all crypt_xyz calls in between should restart the whole process if in the middle of everything one of the crypt_xyz calls results in ENODEV. 

Also the auto-remove DM flag (see dm_deferred_remove()) would have to be turned off as we start using an existing device, so that it doesn#t vanish in the middle of our work, and only be set again when we finished mounting it.

(Speaking of dm_deferred_remove(), we currently call the ioctl directly for that, because libcryptsetup didn't have a native API call for that. Maybe it has now, in which case we should start using that)

Reply:
Ok, sounds good, will work on that. Yes, there is a native API now: crypt_deactivate_by_name with CRYPT_DEACTIVATE_DEFERRED as a flag, will change to use it as well.

Reply:
Funnily enough, there's no interface for _cancelling_ a deferred deactivate flag, I have to call into libdevmapper directly.

Reply:
please don't, libdevmapper is horrible internally, I'd rather not interface with it directly, and add yet another dependency to our build system. Just call the ioctl directly, as the current code already shows how to do.

Reply:
Yeah, it's not great. It's slightly different as it's not a DM_DEV ioctl, it's a DM_TARGET_MSG, but it should not be too different - I'll work on that and push again.

Reply:
https://github.com/torvalds/linux/blob/v5.7/drivers/md/dm-ioctl.c#L1553

Reply:
I've implemented this, with additional logic to fall back to unique/unshared in case libdevmapper does not collaborate. On my machine, with the fixed libcrytpsetup, about one in ten tests of concurrent activations end up with libdevmapper saying it's all good, but then there's no device to mount. With the fallback logic, we do a best-effort optimization and fall back to the current behaviour if the stars are not aligned.
Without the fix in libcryptsetup, these failures in concurrent activations is more common, and what's worse it happens with an unrelated return code -EINVAL (libcryptsetup just passes it through, it's libdevmapper that returns it), so the fallback in even more useful.

Reply:
sent a PR to implement an API to cancel a deferred removal, I'll follow up if it gets accepted: https://gitlab.com/cryptsetup/cryptsetup/-/merge_requests/97